File: data-structure/heap.js
// min heap
class Heap {
    getParent(i) {
        return Math.floor(i / 2);
    }
    getLeftChild(i) {
        return 2 * i;
    }
    getRightChild(i) {
        return 2 * i + 1;
    }
    isLeaf(i) {
        return this.getLeftChild(i) > this.nodes.length;
    }
    show() {
        console.log(this.nodes);
    }
    // return min
    top() {
        return this.nodes[0];
    }
    // return min and remove min
    pop() {
        const min = this.nodes[0];
        this.nodes[0] = this.nodes[this.nodes.length - 1];
        this.nodes.pop();
        let i = 1;
        while(!this.isLeaf(i)) {
            const left = this.getLeftChild(i);
            const right = this.getRightChild(i);
            // swap with the smallest one
            let si = (this.nodes[right - 1] < this.nodes[left - 1]) ?
                right : left;
            if (this.nodes[i - 1] <= this.nodes[si - 1]) {
                // parent is smaller than smallest child
                break;
            }
            const t = this.nodes[i - 1];
            this.nodes[i - 1] = this.nodes[si - 1];
            this.nodes[si - 1] = t;
            i = si;
        }
        return min;
    }
    insert(n) {
        this.nodes.push(n);
        let i = this.nodes.length;
        let parentI = this.getParent(i);
        while (i !== 0 && this.nodes[i - 1] < this.nodes[parentI - 1]) {
            const t = this.nodes[i - 1];
            this.nodes[i - 1] = this.nodes[parentI - 1];
            this.nodes[parentI - 1] = t;
            i = parentI;
            parentI = this.getParent(parentI);
        }
    }
    constructor() {
        // parent: i, starts from 1
        // left child: 2i
        // right child: 2i + 1
        this.nodes = [];
    }
}

const heap = new Heap();
heap.insert(30);
heap.insert(26);
heap.insert(15);
heap.insert(10);
heap.insert(25);
heap.show();
console.log(heap.pop());
heap.show();
console.log(heap.pop());
heap.show();
console.log(heap.pop());
heap.show();

File: data-structure/heap.py
class Heap:
    def getParent(self, i):
        return i // 2
    def getLeft(self, i):
        c = 2 * i
        return c if c < len(self.storage) else -1
    def getRight(self, i):
        c = 2 * i + 1
        return c if c < len(self.storage) else -1
    def swap(self, i, j):
        t = self.storage[i]
        self.storage[i] = self.storage[j]
        self.storage[j] = t
    
    def show(self):
        print(self.storage)

    def pop(self):
        if len(self.storage) == 1: return None
        elif len(self.storage) == 2: return self.storage.pop()
        val = self.storage[1]
        self.storage[1] = self.storage.pop()
        now = 1
        lc = self.getLeft(now)
        rc = self.getRight(now)
        while not (lc == -1 and rc == -1):
            if lc == -1:
                bc = rc       
            elif rc == -1:
                bc = lc
            else:
                bc = lc if self.storage[lc] > self.storage[rc] else rc
            if self.storage[bc] < self.storage[now]:
                break
            self.swap(bc, now)
            now = bc
            lc = self.getLeft(now)
            rc = self.getRight(now)
        return val
        
    def push(self, val):
        self.storage.append(val)
        now = len(self.storage) - 1
        parent = self.getParent(now)
        while (not now == 1) and (self.storage[now] > self.storage[parent]):
            self.swap(now, parent)
            now = parent
            parent = self.getParent(now)
            
    def __init__(self):
        self.storage = [-1]

heap = Heap()
heap.push(10)
heap.push(15)
heap.push(26)
heap.push(30)
heap.push(25)
heap.show()
print(heap.pop())
heap.show()
print(heap.pop())
heap.show()
print(heap.pop())
heap.show()